naga/src/front/atomic_upgrade.rs

```diff
@@ -97,64 +101,72 @@ impl UpgradeState<'_> {
+    /// Otherwise, construct a new one and return that handle.
+    ///
+    /// If `ty` is a [`Pointer`], [`Array`], [`BindingArray`], recurse into the
+    /// type and upgrade its leaf types.
+    ///
+    /// If `ty` is a [`Struct`], recurse into it and upgrade only those fields
+    /// whose indices appear in `field_indices`.
+    ///
+    /// The existing type is not affected.
+    ///
+    /// [`Scalar`]: crate::TypeInner::Scalar
+    /// [`Atomic`]: crate::TypeInner::Atomic
+    /// [`Pointer`]: crate::TypeInner::Pointer
+    /// [`Array`]: crate::TypeInner::Array
+    /// [`Struct`]: crate::TypeInner::Struct
+    /// [`BindingArray`]: crate::TypeInner::BindingArray
+    fn upgrade_type(
+        &mut self,
+        ty: Handle<Type>,
+        field_indices: &mut Vec<u32>,
+    ) -> Result<Handle<Type>, Error> {
         let padding = self.inc_padding();
-        padding.trace("upgrading type: ", ty);
+        padding.trace("visiting type: ", ty);
 
         let inner = match self.module.types[ty].inner {
             TypeInner::Scalar(scalar) => {
                 log::trace!("{padding}hit the scalar leaf, replacing with an atomic");
                 TypeInner::Atomic(scalar)
             }
             TypeInner::Pointer { base, space } => TypeInner::Pointer {
-                base: self.upgrade_type(base)?,
+                base: self.upgrade_type(base, field_indices)?,
                 space,
             },
             TypeInner::Array { base, size, stride } => TypeInner::Array {
-                base: self.upgrade_type(base)?,
+                base: self.upgrade_type(base, field_indices)?,
                 size,
                 stride,
             },
             TypeInner::Struct { ref members, span } => {
-                // In the future we should have to figure out which member needs
-                // upgrading, but for now we'll only cover the single-member
-                // case.
-                let &[crate::StructMember {
-                    ref name,
-                    ty,
-                    ref binding,
-                    offset,
-                }] = &members[..]
-                else {
-                    return Err(Error::MultiMemberStruct);
-                };
-
-                // Take our own clones of these values now, so that
-                // `upgrade_type` can mutate the module.
-                let name = name.clone();
-                let binding = binding.clone();
-                let upgraded_member_type = self.upgrade_type(ty)?;
+                let index = field_indices.pop().ok_or(Error::UnexpectedEndOfIndices)?;
+                let mut new_members = vec![];
+                for (i, mut member) in members.clone().into_iter().enumerate() {
+                    if i == index as usize {
+                        member.ty = self.upgrade_type(member.ty, field_indices)?;
+                    }
+                    new_members.push(member);
+                }
```

Comment:
This seems much more complex than necessary. Isn't this equivalent?

```suggestion
                let index = field_indices.pop().ok_or(Error::UnexpectedEndOfIndices)? as usize;
                let mut new_members = members.clone();
                new_members[index].ty = self.upgrade_type(new_members[index].ty, field_indices)?;
```